src/services/utxoAtPoint.ts

```diff
@@ -0,0 +1,99 @@
+import { Pool } from "pg";
+import { Request, Response } from "express";
+
+import config from "config";
+import {assertNever, validateAddressesReq, getAddressesByType, extractAssets,} from "../utils";
+
+import { getBlock, getLatestBlock } from "../utils/queries/block";
+
+const addressesRequestLimit:number = config.get("server.addressRequestLimit");
+
+const utxoAtPointQuery = `SELECT tx_out.address,
+       tx_out.payment_cred,
+       encode(tx.hash::bytea, 'hex'::text) AS hash,
+       tx_out.index,
+       tx_out.value,
+       block.block_no AS "blockNumber",
+       (
+           SELECT json_agg(
+               ROW (
+                   encode(ma_tx_out.policy::bytea, 'hex'::text),
+                   encode(ma_tx_out.name::bytea, 'hex'::text),
+                   ma_tx_out.quantity)
+               ) AS json_agg
+            FROM ma_tx_out
+            WHERE ma_tx_out.tx_out_id = tx_out.id
+        ) AS assets
+FROM tx
+    JOIN tx_out ON tx.id = tx_out.tx_id
+    JOIN block ON block.id = tx.block_id
+WHERE tx.valid_contract
+    AND block.block_no <= ($3)::uinteger
+    AND NOT utxo_used_as_invalid_collateral(tx_out.tx_id, tx_out.index::smallint, ($3)::uinteger)
+    AND NOT utxo_used_as_valid_input(tx_out.tx_id, tx_out.index::smallint, ($3)::uinteger)
```

Comment:
Since the case for null reference block will be removed, I think there might be an option to get rid of the functions and just join the inputs and colleteral inputs to the same selection

Reply:
The point of the functions is just so the whole query gets a little bit easier to read. I think it's nicer to read through this types of conditions than subqueries

Reply:
No problem, sounds good to me, and I agree it reads nice, let's then just check the execution plan for the query before merging it in to make sure the database resolved properly that a lot of similar tables gets joined within the function and in the query where the function is used 👌

Reply:
Cool, I will do it as well 👍